File: lib/download/downloader.py
import time
from ..talker import Talker
from ..driver.support.driver_support import DriverSupport


class Downloader:
    """Base class for Downloader object"""

    def __init__(self, driver, opts, nav_info, download_filter_obj=None):
        """Initializes Downloader object

        :param WebDriver driver: Selenium WebDriver
        :param Opts opts: Opts class
        :param NavInfo nav_info: NavInfo class
        :param download_filter_obj: download filter to use
        """
        self._driver = driver
        self._opts = opts
        self._nav_info = nav_info
        self._filter_obj = download_filter_obj() if download_filter_obj else None

    def download(self) -> None:
        """Download files located in current directory

        :return:
        """
        elements = self.get_download_elements(self._driver, self._opts)
        if not elements:
            return
        if self._filter_obj:
            elements = self._filter_obj.apply(elements)

        Talker.loading("Downloading Files", new_line=True)

        if self._nav_info.extra_task:
            self._multiple_tasks(elements)
        else:
            self._single_task(elements)

    def right_click_download(self) -> None:
        """Download files using the right click's context menu"""
        elements = DriverSupport.get_elements_all(self._driver,
                                                  self._opts,
                                                  self._nav_info.css_select,
                                                  f"Download Elements "
                                                  f"cannot be found at "
                                                  f"{self._driver.current_url}")
        if self._filter_obj:
            elements = self._filter_obj.apply(elements)
        context_download_element = DriverSupport.get_element(self._driver,
                                                             self._nav_info.css_download)

        for el in elements:
            if self._opts.scroll:
                DriverSupport.scroll_to_element(self._driver, el)
                if self._opts.scroll_wait:
                    time.sleep(self._opts.scroll_wait)
            DriverSupport.right_click(self._driver, el)
            context_download_element.click()

    def get_download_elements(self, driver, opts) -> list:
        """Retrieve all elements needed for downloading

        :return: list of elements related to downloading
        """
        opts.get_wait = lambda: 30
        return DriverSupport.get_elements_wait(driver, opts,
                                               self._nav_info.css_download,
                                               f"Download element cannot be found at"
                                               f"{driver.current_url}!")

    def _single_task(self, elements) -> None:
        """Download operation with a single action

        :param list elements: list of elements
        :return:
        """
        for i in range(len(elements)):
            Downloader._waiting(self._opts)
            elems = self.get_download_elements(self._driver, self._opts)
            if self._filter_obj:
                elems = self._filter_obj.apply(elems)
            elems[i].click()

    def _multiple_tasks(self, elements) -> None:
        """Download operation with multiple actions

        :param list elements: list of elements
        :return:
        """
        for i, el in enumerate(elements):
            Downloader._waiting(self._opts)
            if i == 0:
                el.click()
            else:
                els = self.get_download_elements(self._driver, self._opts)
                if self._filter_obj:
                    els = self._filter_obj.apply(els)
                els[i].click()
            self._nav_info.extra_task(self._driver)

    def multi_task_lazy(self, driver, opts, index):
        Downloader._waiting(opts)
        els = self.get_download_elements(driver, opts)
        if self._filter_obj:
            els = self._filter_obj.apply(els)
        els[index].click()
        self._nav_info.extra_task(driver)

    @staticmethod
    def _waiting(opts) -> None:
        wait = opts.get_download_wait()
        if wait:
            if opts.verbose:
                Talker.loading(f"Waiting for {wait} seconds")
            time.sleep(wait)

File: app/src/main/java/xyz/zimtools/zyod/download/Downloader.java
package xyz.zimtools.zyod.download;

import org.openqa.selenium.WebElement;
import org.openqa.selenium.remote.RemoteWebDriver;
import xyz.zimtools.zyod.AppConfig;
import xyz.zimtools.zyod.args.Args;
import xyz.zimtools.zyod.assets.info.DownloadInfo;
import xyz.zimtools.zyod.assets.info.NavInfo;
import xyz.zimtools.zyod.download.filters.DownloadFilter;
import xyz.zimtools.zyod.support.InteractSupport;
import xyz.zimtools.zyod.support.NavSupport;

import java.util.List;
import java.util.Optional;

/**
 * Downloads resources from an OD
 */
public final class Downloader {
    private final RemoteWebDriver driver;
    private final Args args;
    private final NavInfo navInfo;
    private final DownloadInfo downloadInfo;
    private final DownloadFilter filter;

    public Downloader(RemoteWebDriver driver, Args args, NavInfo navInfo,
                      DownloadInfo downloadInfo, DownloadFilter filter) {
        this.driver = driver;
        this.args = args;
        this.navInfo = navInfo;
        this.downloadInfo = downloadInfo;
        this.filter = filter;
    }

    public Downloader(RemoteWebDriver driver, Args args, NavInfo navInfo, DownloadInfo downloadInfo) {
        this(driver, args, navInfo, downloadInfo, null);
    }

    /**
     * Download files in the current URL.
     */
    private void download(List<WebElement> elements) {
        if (this.downloadInfo.getExtraTasks() != null) {
            this.multipleTasks(elements);
        } else {
            this.singleTask(elements);
        }
    }

    /**
     * Download the first located file.
     */
    public void singleDownload() {
        Optional<WebElement> element = this.getDownloadElement(this.driver);
        if (element.isEmpty()) {
            return;
        }
        List<WebElement> elList = List.of(element.get());
        this.download(elList);
    }

    /**
     * Download multiple files
     */
    public void multiDownload() {
        List<WebElement> elements = this.getDownloadElements(driver);
        if (elements.isEmpty()) {
            return;
        }

        if (this.filter != null) {
            elements = this.filter.apply(elements, this.downloadInfo);
        }
        this.download(elements);
    }

    /**
     * Downloads the first file using the context menu.
     */
    public void rightClickSingleDownload() {
        List<WebElement> elements = NavSupport.getElements(this.driver, this.navInfo.getCssFileSelector());
        if (elements.isEmpty()) {
            return;
        }
        if (this.filter != null) {
            elements = this.filter.apply(elements, this.downloadInfo);
        }

        if (elements.isEmpty()) {
            return;
        }

        WebElement validEl = elements.get(0);
        WebElement contextEl = this.getDownloadElements(driver).get(0);
        this.rightClick(contextEl, validEl);
    }

    /**
     * Downloads multiple files using the context menu.
     */
    public void rightClickMultiDownload() {
        List<WebElement> elements = NavSupport.getElements(this.driver, this.navInfo.getCssFileSelector());
        if (elements.isEmpty()) {
            return;
        }
        if (this.filter != null) {
            elements = this.filter.apply(elements, this.downloadInfo);
        }

        WebElement contextEl = this.getDownloadElements(driver).get(0);
        for (WebElement el : elements) {
            this.rightClick(contextEl, el);
        }
    }

    /**
     * Right-Click to download file
     */
    private void rightClick(WebElement contextEl, WebElement el) {
        if (this.args.getArgsInteractive().isScrolling()) {
            InteractSupport.scrollToElement(this.driver, el);
            long scrollWait = this.args.getArgsInteractive().getScrollWait();
            if (scrollWait > 0) {
                AppConfig.sleep(scrollWait);
            }
        }
        this.waiting();
        InteractSupport.rightClick(driver, el);
        contextEl.click();
    }


    public List<WebElement> getDownloadElements(RemoteWebDriver driver) {
        return NavSupport.getElements(driver, this.downloadInfo.getCssInitialDownload());
    }

    public Optional<WebElement> getDownloadElement(RemoteWebDriver driver) {
        return NavSupport.getElement(driver, this.downloadInfo.getCssInitialDownload());
    }

    /**
     * Download operations with a single action.
     *
     * @param elements list of {@link WebElement} to download.
     */
    private void singleTask(List<WebElement> elements) {
        for (int i = 0; i < elements.size(); i++) {
            this.waiting();
            List<WebElement> downloadEls = this.getDownloadElements(driver);
            if (this.filter != null) {
                downloadEls = this.filter.apply(downloadEls, this.downloadInfo);
            }
            downloadEls.get(i).click();
        }
    }

    /**
     * Downloading operations with multiple actions.
     *
     * @param elements list of {@link WebElement} to download.
     */
    private void multipleTasks(List<WebElement> elements) {
        for (int i = 0; i < elements.size(); i++) {
            this.waiting();
            if (i == 0) {
                elements.get(i).click();
            } else {
                List<WebElement> downloadEls = this.getDownloadElements(driver);
                if (this.filter != null) {
                    downloadEls = this.filter.apply(downloadEls, this.downloadInfo);
                }
                downloadEls.get(i).click();
            }
            this.downloadInfo.getExtraTasks().accept(this.driver);
        }
    }

    public void multipleTaskLazy(int index) {
        this.waiting();
        List<WebElement> downloadEls = this.getDownloadElements(this.driver);
        if (this.filter != null) {
            downloadEls = this.filter.apply(downloadEls, this.downloadInfo);
        }
        downloadEls.get(index).click();
        this.downloadInfo.getExtraTasks().accept(this.driver);
    }

    /**
     * Wait for a certain amount of time before performing a download action.
     * <p>
     * The action this method refers to is downloading. Wait a certain of time before
     * executing downloading task.
     * </p>
     */
    private void waiting() {
        long wait = this.args.getArgsDownload().getDownloadWait();
        if (wait > 0) {
            AppConfig.sleep(wait);
        }
    }


}